blackjax/kernels.py

```diff
@@ -623,6 +624,48 @@ def step_fn(rng_key: PRNGKey, state, minibatch: PyTree, step_size: float):
         return step_fn
 
 
+class csgld:
+
+    init = staticmethod(sgmcmc.csgld.init)
+    kernel = staticmethod(sgmcmc.csgld.kernel)
+
+    def __new__(  # type: ignore[misc]
+        cls,
+        logprob_estimator_fn: Callable,
+        grad_estimator_fn: Callable,
+        zeta: float = 1,
+        temperature: float = 0.01,
+        num_partitions: int = 512,
+        energy_gap: float = 100,
+    ) -> MCMCSamplingAlgorithm:
+
+        step = cls.kernel(num_partitions, energy_gap)
```

Comment:
And then pass it here

Reply:
Done. Could you update the file so that I can continue my test?

Reply:
Yes you should be able to if you installed the library with `pip -e .`

Reply:
> Yes you should be able to if you installed the library with `pip install -e .` in the environment where you're executing your example.

Great, probably my last question: do you know how to update new_state.position? CSGLD becomes so explorative, it may explore the extreme tails and lead to nan values. To avoid it, I have to limit the particle in a bounded domain such that  whenever some extreme values occur, we reset new_state.position to some nice values 

<img width="1600" alt="Screen Shot 2022-12-12 at 12 13 06 PM" src="https://user-images.githubusercontent.com/18079879/207109843-9a1ceb16-0a28-402b-a693-030ce919f424.png">

Reply:
> Yes you should be able to if you installed the library with `pip install -e .` in the environment where you're executing your example.

<img width="1597" alt="Screen Shot 2022-12-12 at 12 14 55 PM" src="https://user-images.githubusercontent.com/18079879/207110189-5315d479-18cd-49ac-925d-02548ad99a9e.png">

Reply:
Try something like

```python 
new_state = CSGLDState(..., state.energy_pdf, state.energy_idx)
```

Reply:
It is just because CSGLD is sooooo great at exploration and is such an expert in avoiding local optimal. 

Imagine that the standard normal distribution, CSGLD can easily explore tails such as 100. the probability or other relevant tools in this value is nan.

Reply:
> 

We can consider to solve issue in future updates. Let me show a perfect demo first.

Reply:
> Try something like
> 
> ```python
> new_state = CSGLDState(4, state.energy_pdf, state.energy_idx)
> ```
> 
> But if you're afraid of getting `NaN` we should probably handle this in a nicer way within the library. What would be the source of those `Nan` values?

Great News! I have implemented the Contour SGLD algorithm and it works perfectly.

<img width="1099" alt="image" src="https://user-images.githubusercontent.com/18079879/207187263-a64710d4-f0e6-485b-9d5b-7ef7d781e8d3.png">

Reply:
> Try something like
> 
> ```python
> new_state = CSGLDState(4, state.energy_pdf, state.energy_idx)
> ```
> 
> But if you're afraid of getting `NaN` we should probably handle this in a nicer way within the library. What would be the source of those `Nan` values?

One thing is that I cannot write md in my side because I am using a server that doesn't allow jupyter notebook.

Is it appropriate to ask you to write a md file? I will prepare a python file that is as close to the md file as possible to minimize your cost.

Reply:
Here is the example python code to reproduce the perfect example: https://github.com/WayneDW/Jax_implementation_of_CSGLD/blob/main/howto_use_csgld_final.py


The generated figures are all presented in the github repo: https://github.com/WayneDW/Jax_implementation_of_CSGLD.git

Reply:
You can take a look now.